src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -0,0 +1,1414 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.Semantics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using System;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    partial class BoundExpression : IExpression
+    {
+        ITypeSymbol IExpression.ResultType => this.Type;
+
+        OperationKind IOperation.Kind => this.ExpressionKind;
+       
+        object IExpression.ConstantValue => this.ConstantValue?.Value;
+
+        SyntaxNode IOperation.Syntax => this.Syntax;
+        
+        protected virtual OperationKind ExpressionKind => OperationKind.None;
+        // protected abstract OperationKind ExpressionKind { get; }
+    }
+
+    internal abstract partial class BoundNode : IOperationSearchable
+    {
+        IEnumerable<IOperation> IOperationSearchable.Descendants()
+        {
+            var list = new List<BoundNode>();
+            new Collector(list).Visit(this);
+            list.RemoveAt(0);
+            return list.OfType<IOperation>();
+        }
+
+        IEnumerable<IOperation> IOperationSearchable.DescendantsAndSelf()
+        {
+            var list = new List<BoundNode>();
+            new Collector(list).Visit(this);
+            return list.OfType<IOperation>();
+        }
+
+        private class Collector : BoundTreeWalker
+        {
+            private readonly List<BoundNode> nodes;
+
+            public Collector(List<BoundNode> nodes)
+            {
+                this.nodes = nodes;
+            }
+
+            public override BoundNode Visit(BoundNode node)
+            {
+                this.nodes.Add(node);
+                return base.Visit(node);
+            }
+        }
+    }
+
+    partial class BoundCall : IInvocation
+    {
+        IMethodSymbol IInvocation.TargetMethod => this.Method;
+
+        IExpression IInvocation.Instance => this.ReceiverOpt;
+       
+        InvocationKind IInvocation.InvocationKind
+        {
+            get
+            {
+                IMethodSymbol method = this.Method;
+
+                if (method.IsStatic)
+                {
+                    return InvocationKind.Static;
+                }
+
+                if ((method.IsVirtual || method.IsAbstract || method.IsOverride) && !this.ReceiverOpt.SuppressVirtualCalls)
```

Comment:
Because the IInvocation already exposes the method, is there value in exposing this InvocationKind flag?  All this can be gotten from the Method and ReceiverOpt.  So should we really have this baked in?